test/e2e/apps/statefulset.go

```diff
@@ -248,6 +248,123 @@ var _ = SIGDescribe("StatefulSet", func() {
+			ss, pods = sst.WaitForPodNotReady(ss, pods.Items[1].Name)
+			newImage := NewNginxImage
+			oldImage := ss.Spec.Template.Spec.Containers[0].Image
+
+			By(fmt.Sprintf("Updating StatefulSet template: update image from %s to %s", oldImage, newImage))
+			Expect(oldImage).NotTo(Equal(newImage), "Incorrect test setup: should update to a different image")
+			ss, err = framework.UpdateStatefulSetWithRetries(c, ns, ss.Name, func(update *apps.StatefulSet) {
+				update.Spec.Template.Spec.Containers[0].Image = newImage
+			})
+			Expect(err).NotTo(HaveOccurred())
+
+			By("Creating a new revision")
+			ss = sst.WaitForStatus(ss)
+			currentRevision, updateRevision = ss.Status.CurrentRevision, ss.Status.UpdateRevision
+			Expect(currentRevision).NotTo(Equal(updateRevision),
+				"Current revision should not equal update revision during rolling update")
+
+			By("Updating Pods in reverse ordinal order")
+			pods = sst.GetPodList(ss)
+			sst.SortStatefulPods(pods)
+			sst.RestorePodHttpProbe(ss, &pods.Items[1])
+			ss, pods = sst.WaitForPodReady(ss, pods.Items[1].Name)
+			ss, pods = sst.WaitForRollingUpdate(ss)
+			Expect(ss.Status.CurrentRevision).To(Equal(updateRevision),
+				fmt.Sprintf("StatefulSet %s/%s current revision %s does not equal update revision %s on update completion",
+					ss.Namespace,
+					ss.Name,
+					ss.Status.CurrentRevision,
+					updateRevision))
+			for i := range pods.Items {
+				Expect(pods.Items[i].Spec.Containers[0].Image).To(Equal(newImage),
+					fmt.Sprintf(" Pod %s/%s has image %s not have new image %s",
+						pods.Items[i].Namespace,
+						pods.Items[i].Name,
+						pods.Items[i].Spec.Containers[0].Image,
+						newImage))
+				Expect(pods.Items[i].Labels[apps.StatefulSetRevisionLabel]).To(Equal(updateRevision),
+					fmt.Sprintf("Pod %s/%s revision %s is not equal to update revision %s",
+						pods.Items[i].Namespace,
+						pods.Items[i].Name,
+						pods.Items[i].Labels[apps.StatefulSetRevisionLabel],
+						updateRevision))
+			}
+
+			By("Rolling back to a previous revision")
+			sst.BreakPodHttpProbe(ss, &pods.Items[1])
+			Expect(err).NotTo(HaveOccurred())
+			ss, pods = sst.WaitForPodNotReady(ss, pods.Items[1].Name)
+			priorRevision := currentRevision
+			currentRevision, updateRevision = ss.Status.CurrentRevision, ss.Status.UpdateRevision
+			ss, err = framework.UpdateStatefulSetWithRetries(c, ns, ss.Name, func(update *apps.StatefulSet) {
+				update.Spec.Template.Spec.Containers[0].Image = oldImage
+			})
+			Expect(err).NotTo(HaveOccurred())
+			ss = sst.WaitForStatus(ss)
+			currentRevision, updateRevision = ss.Status.CurrentRevision, ss.Status.UpdateRevision
+			Expect(currentRevision).NotTo(Equal(updateRevision),
+				"Current revision should not equal update revision during roll back")
+			Expect(priorRevision).To(Equal(updateRevision),
+				"Prior revision should equal update revision during roll back")
+
+			By("Rolling back update in reverse ordinal order")
```

Comment:
How are you testing rollback ? Statefulsets doesn’t support it or did you mean rolling update ?

Reply:
This is mostly a copy of the test starting [here](https://github.com/kubernetes/kubernetes/pull/64350/files/cf7ddbe837d2baf1918230c641672c4e6ae70134#diff-cb00e9492a5e88ff0e10e4b82878d690R373), so I'm just doing what they did. Looks like they update the container image to a previous version and call that a rollback.